### PR TITLE
switch semgrep scan --config <product> to use the same endpoint as semgrep ci

### PR DESCRIPTION
**Pull request by jbergler**
Closes https://linear.app/semgrep/issue/SCRT-243/add-secrets-preset-to-config-option

## Context

We have too many different ways of generating config for the CLI and it makes reliability / consistency really hard. Moving forward the hope is the CLI exclusively uses one endpoint (`POST /api/cli/scans`) for getting config that is integrated with [semgrep.dev](http://semgrep.dev/).  

This PR changes the where we fetch the configuration for the following commands: `--config [secrets/supply-chain/code]`.

## Changes

- Switched `--config <product>` to use same endpoint as semgrep ci (`/api/cli/scans`)
- Added support for `--config secrets`
    - When using
code or secrets, the environment variable `SEMGREP_REPO_NAME` must be set.
- Fixed issue where secrets validations were not being run even though they were downloaded from a trusted repository.

## Future considerations

* The fallback url still uses query params rather than the POST body used for the new endpoint. This means we still need to generate the old config url. 

## Local run (excerpt)

```
$ SEMGREP_REPO_NAME=test SEMGREP_APP_TOKEN=token_here semgrep scan --config secrets
               
               
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 6 files (only git-tracked) with 0 Code rules, 267 Secrets rules:
            
  CODE RULES
  Nothing to scan.
               
  SECRETS RULES
                                                                                                                        
  Language      Rules   Files          Origin      Rules                                                                
 ─────────────────────────────        ───────────────────                                                               
  <multilang>     211      10          Pro rules     267                                                                
  python           10       1                                                                                           
  json              2       1                                                                                           
                                                                                                                        
                    
  SUPPLY CHAIN RULES
                                                       
  💎 Run `semgrep ci` to find dependency
     vulnerabilities and advanced cross-file findings. 
                                                       
          
  PROGRESS
   
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:01                                                                                                                        
                           
                           
┌─────────────────────────┐
│ 1 Valid Secrets Finding │
└─────────────────────────┘
...

...

...
```

### CODE REVIEW

**Comment by aryx** [comment]
Do we want that? Do we want semgrep scan --config <product>?
What's the difference with 'semgrep ci' then?
cc @milanwilliams @zzeleznick 
I think there is some overlap between semgrep ci and semgrep scan and I'm not sure it's a good thing.

File: cli/src/semgrep/config_resolver.py [hunk 1]
```diff
@@ -805,6 +898,19 @@ def url_for_policy() -> str:
     return f"{env.semgrep_url}/{DEFAULT_SEMGREP_APP_CONFIG_URL}?{params_str}"
 
 
+PRODUCT_NAMES = {
```
Comment (zzeleznick) [inline cli/src/semgrep/config_resolver.py:952]: The developer ergonomics for interacting with the set of products we supports seems rusty.

#### `rule_match.py`
```python
# TODO: diff with rule.py product() method?
@property
def product(self) -> out.Product:
    if self.metadata.get("product") == "secrets":
        return out.Product(out.Secrets())
    elif "sca_info" in self.extra:
        return out.Product(out.SCA())
    else:
        return out.Product(out.SAST())
```

#### `rule.py`
```python
@property
def product(self) -> out.Product:
    if "r2c-internal-project-depends-on" in self._raw:
        return out.Product(out.SCA())
    elif "product" in self.metadata:
        if self.metadata.get("product") == "secrets":
            return out.Product(out.Secrets())
        else:
            return out.Product(out.SAST())
    else:
        return out.Product(out.SAST())
```

**Comment by jbergler** [comment]
> Do we want that? Do we want semgrep scan --config ? What's the difference with 'semgrep ci' then? cc @milanwilliams @zzeleznick I think there is some overlap between semgrep ci and semgrep scan and I'm not sure it's a good thing.

I believe we do want this since each of the product teams has been implementing this - first was `--config policy` from ages ago, then supply chain added `--config supply-chain` and now theres a secrets scope doc to add `--secrets`

The context from previous discussions that @milanwilliams has driven (afaik) is something like
- semgrep scan -> doesn't require git, doesn't do diff scans, lightly integrated with [semgrep.dev](http://semgrep.dev/) (can use the configs from scp)
- semgrep ci -> requires git, does diff scans, heavily integrated with [semgrep.dev](http://semgrep.dev/) (findings, triage, etc)

AFAIK `semgrep scan` support for products like supply chain and secrets is a product ask driven by customers who aren't using git.

I put this change together for @salolivares who is working on this for secrets. The motivation to use the same endpoint is to address a class of bugs that exist because we have a whole suite of endpoints that do different things for generating product configs for a logged in user and unifying these makes the experience consistent, and easy to support.

File: cli/src/semgrep/config_resolver.py (hunk 1 above)
Comment (jbergler) [reply cli/src/semgrep/config_resolver.py:952]: Yeah, 100% theres opportunity to improve this.
Some useful context on this PR is that I threw this together in 30min for @salolivares to pick up and run with. I shared a bunch of notes with him about things that need improving and we can totally add this to the list (although it does feel a bit like scope creep)

File: cli/src/semgrep/config_resolver.py [hunk 2]
```diff
@@ -253,6 +261,91 @@ def _load_config_from_local_path(self) -> List[ConfigFile]:
     def is_registry_url(self) -> bool:
         return self._origin == ConfigType.REGISTRY
 
+    def _project_metadata_for_standalone_scan(self) -> out.ProjectMetadata:
+        repo_name = os.environ.get("SEMGREP_REPO_NAME")
+        if repo_name is None:
```
Comment (salolivares) [inline cli/src/semgrep/config_resolver.py:264]: @jbergler How necessary is the `repo_name`? This check no longer makes it possible to run `semgrep scan --config supply-chain` (supply chain was working before, and same applies to secrets). 

If we're requiring it be defined, can we pass "unknown"? or figure out a different way to grab repo name?

Comment (salolivares) [reply cli/src/semgrep/config_resolver.py:264]: Looks like `SEMGREP_REPO_NAME` is [semgrep ci specific](https://semgrep.dev/docs/semgrep-ci/ci-environment-variables/#semgrep_repo_name). Does it make sense to skip this check (or is ConfigLoader used for `ci` as well in which case we skip the check for `scan` only)?

Comment (zzeleznick) [reply cli/src/semgrep/config_resolver.py:264]: might take a peek at `GitMeta` as another flavor of pulling `repo_name` using `git rev-parse` – 
https://github.com/semgrep/semgrep/blob/b47ebc9c0a914b669addc069e0160b3f5193a80c/cli/src/semgrep/meta.py#L103

File: cli/src/semgrep/commands/scan.py [hunk 3]
```diff
@@ -454,6 +454,11 @@ def scan(
 
             version_check()
         return None
+    
+    if config and "secrets" in config:
```
Comment (salolivares) [inline cli/src/semgrep/commands/scan.py:461]: @jonas I wish there was an easy way to leverage the `engine_params` from the new `GET /api/cli/scans` endpoint here but that info is not available until we fetch the rules which happens further along when fetching the config. 

This was the easiest solution.

File: cli/src/semgrep/config_resolver.py (hunk 1 above)
Comment (salolivares) [reply cli/src/semgrep/config_resolver.py:952]: Going to defer this: don't have a good enough understanding of the CLI yet + don't wanna balloon the scope of this PR. Is there a good linear project I can throw this improvement in?

File: cli/src/semgrep/commands/scan.py (hunk 3 above)
Comment (jbergler) [reply cli/src/semgrep/commands/scan.py:461]: Probably worth leaving this as a comment in the code for our future selves.

File: cli/src/semgrep/config_resolver.py (hunk 2 above)
Comment (jbergler) [reply cli/src/semgrep/config_resolver.py:264]: `semgrep scan` can't have git dependencies as it stands, so I'd rather just limit requiring the env var if code or secrets are being requested, and we'll figure out auto detection / adding it for supply chain later.

eg:

```suggestion
        if repo_name is None and require_repo_name:
```

File: cli/src/semgrep/config_resolver.py [hunk 4]
```diff
@@ -250,6 +255,131 @@ def _load_config_from_local_path(self) -> List[ConfigFile]:
...
+
+        request = out.ScanRequest(
+            meta=out.RawJson({}),  # required for now, but we won't populate it
+            scan_metadata=out.ScanMetadata(
+                cli_version=out.Version(__VERSION__),
+                unique_id=out.Uuid(str(state.request_id)),
+                requested_products=products,
+                dry_run=True,  # semgrep scan never submits findings, so always a dry run
+            ),
+            project_metadata=self._project_metadata_for_standalone_scan(),
```
Comment (jbergler) [inline cli/src/semgrep/config_resolver.py]:
```suggestion
            project_metadata=self._project_metadata_for_standalone_scan(require_repo_name),
```

File: cli/src/semgrep/config_resolver.py [hunk 5]
```diff
@@ -250,6 +255,131 @@ def _load_config_from_local_path(self) -> List[ConfigFile]:
...
+    def _fetch_semgrep_cloud_platform_scan_config(self) -> ConfigFile:
+        """
+        Download a configuration from semgrep.dev using new /api/cli/scans endpoint
+        """
+        state = get_state()
+
+        products = [
+            out.Product.from_json(PRODUCT_NAMES[p])
+            for p in self._config_path.split(",")
+        ]
```
Comment (jbergler) [inline cli/src/semgrep/config_resolver.py:302]:
```suggestion
        ]
        require_repo_name = set(products) & {SAST(), Secrets()} 
```

File: cli/tests/unit/test_config_resolver.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        config_loader: ConfigLoader,
+        mocked_state,
+        mocked_cloud_platform_request: out.ScanRequest,
+        mocked_scan_response: out.ScanResponse,
+        mocker,
+    ):
+        mock_response = mocker.MagicMock()
+        mock_response.status_code = 200
+        mock_response.json.return_value = mocked_scan_response.to_json()
+        mocked_state.app_session.post.return_value = mock_response
```
Comment (jbergler) [inline cli/tests/unit/test_config_resolver.py]: It's much better to use the `requests_mock` plugin/fixture to mock HTTP requests than stubbing request objects by hand. Theres loads of examples in `test_ci.py`

If you're open to it, I'd prefer that we take that approach here.

File: cli/src/semgrep/config_resolver.py (hunk 4 above)
Comment (salolivares) [reply cli/src/semgrep/config_resolver.py]: Done in [fc41789](https://github.com/semgrep/semgrep/pull/9205/commits/fc41789b107c847fdf58199b621316a77509f6bc)

File: cli/src/semgrep/config_resolver.py (hunk 5 above)
Comment (salolivares) [reply cli/src/semgrep/config_resolver.py:302]: Done in [fc41789](https://github.com/semgrep/semgrep/pull/9205/commits/fc41789b107c847fdf58199b621316a77509f6bc)

File: cli/src/semgrep/config_resolver.py (hunk 2 above)
Comment (salolivares) [reply cli/src/semgrep/config_resolver.py:264]: Done in [fc41789](https://github.com/semgrep/semgrep/pull/9205/commits/fc41789b107c847fdf58199b621316a77509f6bc). 

`--config <secrets|code>` requires SEMGREP_REPO_NAME.

File: cli/src/semgrep/commands/scan.py (hunk 3 above)
Comment (salolivares) [reply cli/src/semgrep/commands/scan.py:461]: Done in [f70afa6](https://github.com/semgrep/semgrep/pull/9205/commits/f70afa6daf1cfc5be3726595106e1c3ca22dd8ad)

File: cli/tests/unit/test_config_resolver.py (hunk 6 above)
Comment (salolivares) [reply cli/tests/unit/test_config_resolver.py]: Done in [fa09f29](https://github.com/semgrep/semgrep/pull/9205/commits/fa09f298b53944146f936b6f4eed6c69d2dab22a) and [9c920ba](https://github.com/semgrep/semgrep/pull/9205/commits/9c920bad1882e74f5d0e2f83a1d5bb94b75c0732)

**Comment by salolivares** [comment]
@jbergler changelog added in [c9596d9](https://github.com/semgrep/semgrep/pull/9205/commits/c9596d90404c98f33a73d68577f360c98a9ac45d) and addressed all pending feedback! Let me know if I missed anything.

File: cli/src/semgrep/config_resolver.py (hunk 1 above)
Comment (zzeleznick) [reply cli/src/semgrep/config_resolver.py:952]: @salolivares – I can't say for certain how we track eliminating tech-debt from our codebases, and I don't believe we have a label / project for a task like this. Normally I'd encourage creating a linear ticket and marking the future work with a comment linking the ticket; however, this was more a call-out than a request for changes.

Comment (zzeleznick) [reply cli/src/semgrep/config_resolver.py:952]: feel free to resolve this conversation!

**Review by zzeleznick — APPROVED** [review]
LGTM w/ comment –

Not quite sure of the intended final CLI output, but there's coupling in https://github.com/semgrep/semgrep/blob/develop/cli/src/semgrep/state.py#L58 to determine which template pattern to use.

**Review by jbergler — COMMENTED** [review]
LGTM

**Comment by salolivares** [comment]
> Not quite sure of the intended final CLI output, but there's coupling in https://github.com/semgrep/semgrep/blob/develop/cli/src/semgrep/state.py#L58 to determine which template pattern to use.

@zzeleznick created a [linear ticket](https://linear.app/semgrep/issue/SCRT-254/display-detailed-ux-for-semgrep-scan-config-secrets) to address this. thanks for the reminder!

**Pull request merged**